### PR TITLE
feat: add support for postgres schema selection

### PR DESCRIPTION
**Pull request by gkorland**
## Summary

Adds support for selecting a PostgreSQL schema instead of always using `public`. Based on [PR #373](https://github.com/FalkorDB/QueryWeaver/pull/373) by @sirudog, rebased onto current staging with bug fixes.

### What changed
- **Backend**: New `_parse_schema_from_url()` extracts schema from the connection URL's `options` parameter (`search_path`), following PostgreSQL's native libpq format
- **SQL queries**: All `information_schema` queries now accept a parameterized `schema` argument (default: `public`) — no SQL injection risk
- **UI**: Schema input field in DatabaseModal, shown only for PostgreSQL connections
- **Tests**: 14 new unit tests for URL schema parsing (edge cases, encoding, \$user handling)
- **Docs**: Custom schema configuration guide with examples and troubleshooting

### Fixes over original PR #373
1. **`pg_namespace` JOIN order in `extract_columns_info`** — the original used `LEFT JOIN pg_class` then `LEFT JOIN pg_namespace`, which doesn't filter duplicate `pg_class` rows when same-named tables exist across schemas. Fixed by joining `pg_namespace` first and folding the namespace filter into the `pg_class` join condition
2. **Regex pattern** — changed `[=\s]+` to `\s*=\s*` to require `=` as separator, preventing mis-capture when `search_path=` is followed by a space and another option
3. **\$user handling** — replaced two-position check with a loop through all schemas, correctly handling edge cases like `\$user,\$user,public`
4. **Pylint compliance** — fixed line-too-long, added protected-access disable for test class

### Backward compatibility
All methods default to `public` schema, so existing connections without `search_path` work identically.

Closes #373 (supersedes)

## Summary by CodeRabbit

* **New Features**
  * PostgreSQL loader supports selecting a custom schema via connection URL; UI modal adds a PostgreSQL schema input with validation.

* **Documentation**
  * Docs updated with custom schema configuration, examples, verification, and troubleshooting guidance.

* **Bug Fixes**
  * Fixed code-fence formatting in README streaming example.

* **Tests**
  * Added comprehensive unit tests for URL-based schema parsing; marked several e2e chat tests as slow.

* **Chores**
  * Minor repository housekeeping and ignore-list updates; updated public wordlist entries.

### CODE REVIEW

File: app/src/components/modals/DatabaseModal.tsx [hunk 1]
```diff
@@ -94,6 +95,12 @@ const DatabaseModal = ({ open, onOpenChange }: DatabaseModalProps) => {
         const builtUrl = new URL(`${protocol}://${host}:${port}/${database}`);
         builtUrl.username = username;
         builtUrl.password = password;
+        
+        // Append schema option for PostgreSQL if provided
```
[the remark this answers is not in the document]
Comment (gkorland) [reply app/src/components/modals/DatabaseModal.tsx:100]: Fixed in e21496a — the schema input is now sanitized to allow only valid PostgreSQL identifier characters (`[a-zA-Z0-9_]`) before being interpolated into the options string. Invalid characters are stripped.

File: docs/postgres_loader.md [hunk 2]
```diff
@@ -228,8 +257,29 @@ PostgreSQL schema loaded successfully. Found 15 tables.
...
+    GRANT USAGE ON SCHEMA my_schema TO my_user;
+    GRANT SELECT ON ALL TABLES IN SCHEMA my_schema TO my_user;
+    ```
+
+### Verifying Schema Configuration
+
+To verify which schema will be used, you can test the search_path parsing:
+
+{% capture python_2 %}
+from api.loaders.postgres_loader import PostgresLoader
```
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md:276]: This is not an issue — the class is actually named `PostgresLoader` (see `api/loaders/postgres_loader.py` line 70: `class PostgresLoader(BaseLoader)`), which matches what the documentation uses. There is no `PostgreSQLLoader` class in this codebase.

File: api/loaders/postgres_loader.py [hunk 3]
```diff
@@ -96,19 +97,65 @@ def _serialize_value(value):
...
+            options_str = unquote(options[0])
+
+            # Parse -c search_path=value from options
+            # Format can be: -csearch_path=schema or -c search_path=schema
+            match = re.search(r'-c\s*search_path\s*=\s*([^\s]+)', options_str, re.IGNORECASE)
+            if match:
+                search_path = match.group(1)
+                schemas = search_path.split(',')
+                for s in schemas:
+                    s = s.strip().strip('"\'')
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py:132]: This is an extremely rare edge case. PostgreSQL schema names containing commas require double-quoting everywhere they're used, and in practice this almost never occurs. The current parser correctly handles all common patterns (simple names, quoted names, \$user, multiple schemas). Additionally, the frontend now sanitizes schema input to only allow `[a-zA-Z0-9_]` characters, making it impossible to enter a schema name with commas through the UI. Supporting quoted-identifier-aware CSV parsing would add significant complexity for a scenario that is virtually never encountered in real-world usage.

File: api/loaders/postgres_loader.py [hunk 4]
```diff
@@ -239,15 +290,17 @@ def extract_columns_info(cursor, table_name: str) -> Dict[str, Any]:
                 JOIN information_schema.key_column_usage ku
                     ON tc.constraint_name = ku.constraint_name
                 WHERE tc.table_name = %s
+                AND tc.table_schema = %s
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py:311]: Good catch — fixed in e21496a. Added `AND tc.constraint_schema = ku.constraint_schema` to both the PK and FK constraint subquery JOINs in `extract_columns_info`. This prevents cross-schema constraint name collisions when the same constraint name exists in different schemas.

File: tests/test_postgres_loader.py [hunk 5]
```diff
@@ -150,6 +150,95 @@ def test_extract_relationships(self):
...
+    def test_search_path_quoted_schema(self):
+        """Test parsing quoted schema name"""
+        url = "postgresql://user:pass@localhost:5432/mydb?options=-csearch_path%3D%22my_schema%22"
+        result = PostgresLoader._parse_schema_from_url(url)
+        self.assertEqual(result, 'my_schema')
+
+    def test_search_path_case_insensitive(self):
+        """Test that search_path matching is case insensitive"""
+        url = "postgresql://user:pass@localhost:5432/mydb?options=-cSEARCH_PATH%3Dmy_schema"
+        result = PostgresLoader._parse_schema_from_url(url)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply tests/test_postgres_loader.py]: This comment is outdated (as GitHub marks it). The code already handles \$user in any position via a loop. I've added explicit tests in e21496a for repeated \$user entries: `test_search_path_repeated_dollar_user_returns_public` and `test_search_path_repeated_dollar_user_with_schema` to cover this regression case.

File: tests/test_postgres_loader.py [hunk 6]
```diff
@@ -150,6 +150,95 @@ def test_extract_relationships(self):
...
+
+    def test_empty_options_returns_public(self):
+        """Test that URL with empty options returns 'public'"""
+        url = "postgresql://user:pass@localhost:5432/mydb?"
+        result = PostgresLoader._parse_schema_from_url(url)
+        self.assertEqual(result, 'public')
+
+    def test_options_without_search_path_returns_public(self):
+        """Test that options without search_path returns 'public'"""
+        url = "postgresql://user:pass@localhost:5432/mydb?options=-csynchronous_commit%3Dlocal"
```
[the remark this answers is not in the document]
Comment (gkorland) [reply tests/test_postgres_loader.py:170]: Good point — the existing code already handles this correctly (empty strings are falsy in the `if s and s != '$user'` check), but it wasn't tested. Fixed in e21496a with two new tests: `test_search_path_empty_tokens_returns_public` (for `search_path=,`) and `test_search_path_blank_quoted_tokens_returns_public` (for `search_path="",""`). Both correctly return `'public'`.

File: api/loaders/postgres_loader.py [hunk 7]
```diff
@@ -96,19 +97,65 @@ def _serialize_value(value):
...
+
+            options = query_params.get('options', [])
+            if not options:
+                return 'public'
+
+            options_str = unquote(options[0])
+
+            # Parse -c search_path=value from options
+            # Format can be: -csearch_path=schema or -c search_path=schema
+            match = re.search(r'-c\s*search_path\s*=\s*([^\s]+)', options_str, re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py]: Good catch — fixed in 9bc3dee. Updated the regex to `(.+?)(?=\s+-c|\s*$)` as suggested, so values like `$user, public` (with space after comma) are captured correctly. Added a test `test_search_path_dollar_user_space_after_comma` to cover this case.

File: api/loaders/postgres_loader.py [hunk 8]
```diff
@@ -96,19 +97,65 @@ def _serialize_value(value):
...
         Returns:
             Tuple[bool, str]: Success status and message
         """
         try:
+            # Parse schema from connection URL (defaults to 'public')
+            schema = PostgresLoader.parse_schema_from_url(connection_url)
+
             # Connect to PostgreSQL database
             conn = psycopg2.connect(connection_url)
             cursor = conn.cursor()
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py:164]: Good point — fixed in 9bc3dee. Now explicitly setting `SET search_path TO <schema>` (using `psycopg2.sql.Identifier` for safe quoting) right after connecting. This ensures all unqualified table references in sample queries resolve to the correct schema.

File: docs/postgres_loader.md [hunk 9]
```diff
@@ -86,6 +87,34 @@ postgresql://[username[:password]@][host[:port]][/database][?options]
 - `postgresql://user:pass@example.com:5432/production_db`
 - `postgresql://postgres@127.0.0.1/testdb`
 
+### Custom Schema Configuration
+
+By default, the loader extracts tables from the `public` schema. To use a different schema, add the `search_path` option to your connection URL using PostgreSQL's standard `options` parameter. 
+More info [here](https://www.postgresql.org/docs/18/runtime-config-client.html#GUC-SEARCH-PATH).
```
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md]: Fixed in 9bc3dee — updated to `/docs/current/` as suggested.

File: docs/postgres_loader.md [hunk 10]
```diff
@@ -228,8 +257,29 @@ PostgreSQL schema loaded successfully. Found 15 tables.
 
 1. **Connection Failed**: Verify the connection URL format and database credentials
 2. **Permission Denied**: Ensure the database user has read access to system tables
-3. **Schema Not Found**: Check that tables exist in the 'public' schema
+3. **No Tables Found**:
+- Check that tables exist in the target schema
+- If using a custom schema, verify the `search_path` option is correctly formatted
+- Ensure the schema name is spelled correctly (schema names are case-sensitive)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md]: Fixed in 9bc3dee — updated the wording to clarify that schema names are only case-sensitive when created or referenced with double quotes, and that unquoted identifiers are folded to lower-case.

File: api/loaders/postgres_loader.py [hunk 11]
```diff
@@ -96,35 +97,88 @@
...
+            options = query_params.get('options', [])
+            if not options:
+                return 'public'
+
+            options_str = unquote(options[0])
+
+            # Parse -c search_path=value from options
+            # Format can be: -csearch_path=schema or -c search_path=schema
+            # Capture up to the next "-c" option (or end of string) so values like "$user, public" work.
+            match = re.search(r'-c\s*search_path\s*=\s*(.+?)(?=\s+-c|\s*$)', options_str, re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py]: Fixed in 5e31f3c. Replaced the ReDoS-vulnerable `(.+?)(?=\s+-c|\s*$)` pattern with `([^\s,]+(?:\s*,\s*[^\s,]+)*)` which uses unambiguous character classes. Each quantifier consumes disjoint character sets so there's no polynomial backtracking. All 23 tests pass.

File: api/loaders/postgres_loader.py [hunk 12]
```diff
@@ -96,35 +97,88 @@ def _serialize_value(value):
...
+            # Parse -c search_path=value from options
+            # Format can be: -csearch_path=schema or -c search_path=schema
+            # Match comma-separated schema tokens (supports spaces after commas).
+            match = re.search(r'-c\s*search_path\s*=\s*([^\s,]+(?:\s*,\s*[^\s,]+)*)', options_str, re.IGNORECASE)
+            if match:
+                search_path = match.group(1)
+                schemas = search_path.split(',')
+                for s in schemas:
+                    s = s.strip().strip('"\'')
+                    if s and s != '$user':
```
[the remark this answers is not in the document]
Comment (gkorland) [reply api/loaders/postgres_loader.py:133]: This is a deliberate trade-off. The previous regex `(.+?)(?=\s+-c|\s*$)` could capture quoted tokens with spaces, but it had a polynomial ReDoS vulnerability (CodeQL flagged it). The replacement `([^\s,]+(?:\s*,\s*[^\s,]+)*)` uses disjoint character classes to prevent backtracking — at the cost of not supporting quoted identifiers with embedded whitespace.

In practice, schema names with spaces are extremely rare (they require double-quoting everywhere in PostgreSQL). The frontend also validates schema input to only allow `[a-zA-Z0-9_]` characters. Supporting space-in-quotes would require a more complex tokenizer (CSV-aware with quote tracking), which adds significant complexity for an edge case that's virtually never encountered.

File: app/src/components/modals/DatabaseModal.tsx [hunk 13]
```diff
@@ -94,6 +95,15 @@ const DatabaseModal = ({ open, onOpenChange }: DatabaseModalProps) => {
...
+        if (selectedDatabase === 'postgresql' && schema.trim()) {
+          const sanitized = schema.trim().replace(/[^a-zA-Z0-9_]/g, '');
+          if (sanitized) {
+            builtUrl.searchParams.set('options', `-csearch_path=${sanitized}`);
+          }
+        }
```
[the remark this answers is not in the document]
Comment (gkorland) [reply app/src/components/modals/DatabaseModal.tsx:106]: Good point — fixed in a12fa88. Changed from silent sanitization to explicit validation. The schema input now shows a red border and error message when invalid characters are entered, and the connection will throw an error if submitted with invalid characters. This ensures the user sees exactly what's wrong instead of having their input silently mutated.

File: docs/postgres_loader.md [hunk 14]
```diff
@@ -228,8 +257,29 @@ PostgreSQL schema loaded successfully. Found 15 tables.
...
+
+### Verifying Schema Configuration
+
+To verify which schema will be used, you can test the search_path parsing:
+
+{% capture python_2 %}
+from api.loaders.postgres_loader import PostgresLoader
+
+# Test URL parsing
+url = "postgresql://user:pass@localhost:5432/mydatabase?options=-c search_path=my_schema"
```
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md:279]: Fixed in a12fa88 — updated the example URL to use URL-encoded form (`-csearch_path%3Dmy_schema`) as suggested.

File: docs/postgres_loader.md [hunk 15]
```diff
@@ -228,8 +257,29 @@ PostgreSQL schema loaded successfully. Found 15 tables.
...
+schema = PostgresLoader.parse_schema_from_url(url)
```
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md:280]: This is incorrect — the class IS named `PostgresLoader` (see `api/loaders/postgres_loader.py` line 70: `class PostgresLoader(BaseLoader)`). There is no `PostgreSQLLoader` class in this codebase. The documentation correctly references `PostgresLoader`, which matches the implementation. The suggested change to `PostgreSQLLoader` would cause an import error.

File: docs/postgres_loader.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (gkorland) [reply docs/postgres_loader.md:279]: Update: Reverted the URL-encoding change in 6e33932. The URL is inside a Liquid `{% capture %}` block (not a fenced code block), so the spellchecker processes it as raw text and flags `csearch` and `Dmy` as misspellings. Since this is a Python code example, `urlparse`/`parse_qs` handles the readable form `-c search_path=my_schema` correctly. The inline code examples elsewhere in the doc already use URL-encoded forms inside backticks (which the spellchecker ignores).
